File: ts/2022/day13.ts
import { main } from 'lib/advent';
import { Iter, iter } from 'lib/iter';
import { paragraphs } from 'lib/util';

type Packet = number | (Packet | number)[];
type Pair = [Packet, Packet];

function parse(paras: string[][]): Pair[] {
  return paras.map((lines) => {
    const [a, b] = lines;
    return [JSON.parse(a), JSON.parse(b)];
  });
}

function packetOrder(left: Packet, right: Packet): number {
  if (typeof left === 'number' && typeof right === 'number')
    return left - right;

  if (!Array.isArray(left)) left = [left];
  if (!Array.isArray(right)) right = [right];

  const length = Math.min(left.length, right.length);
  for (let i = 0; i < length; ++i) {
    const result = packetOrder(left[i], right[i]);
    if (result !== 0) return result;
  }
  return left.length - right.length;
}

function orderedIndexes(pairs: Pair[]): Iter<number> {
  return iter(pairs)
    .map(([left, right], i) => [packetOrder(left, right), i + 1])
    .filter(([order]) => order < 0)
    .map(([, index]) => index);
}

function decoderKey(pairs: Pair[], a: Packet, b: Packet): number {
  const packets = [a, b, ...pairs.flat(1)];
  packets.sort(packetOrder);
  return (
    (packets.findIndex((x) => x === a) + 1) *
    (packets.findIndex((x) => x === b) + 1)
  );
}

main(
  (s) => orderedIndexes(parse(paragraphs(s))).sum(),
  (s) => decoderKey(parse(paragraphs(s)), [[2]], [[6]])
);

File: py/2022/day13.py
import functools as ft
import itertools as it
import json

from aoc import main
from aoc.parse import paras


def parse(s: str):
    return [(a, b) for a, b in paras(s, json.loads)]


def cmp(a, b):
    if isinstance(a, int) and isinstance(b, int):
        return a - b
    if isinstance(a, int):
        a = [a]
    elif isinstance(b, int):
        b = [b]
    for i in range(min(len(a), len(b))):
        r = cmp(a[i], b[i])
        if r != 0:
            return r
    return len(a) - len(b)


def ordered_indexes(pairs):
    return sum(i + 1 for i, valid in enumerate(cmp(a, b) < 1 for a, b in pairs) if valid)


def decoder_key(pairs):
    dividers = [[[2]], [[6]]]
    packets = list(it.chain(dividers, *pairs))
    packets.sort(key=ft.cmp_to_key(cmp))
    return (packets.index(dividers[0]) + 1) * (packets.index(dividers[1]) + 1)


if __name__ == "__main__":
    main(
        lambda s: ordered_indexes(parse(s)),
        lambda s: decoder_key(parse(s)),
    )
